### PR TITLE
Optimize `StringBuilder` by embracing spans and `ISpanFormattable` more.

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
This PR changes the implementation of `StringBuilder` to completely eliminate pinning by moving all pointer-based implementations to spans.

For example, the `Append` overload that takes a `ReadOnlySpan` of characters used to be implemented by pinning the span and forwarding to the `Append` overload that takes a pointer and a length. Now, the roles are reversed: the pointer overload calls the `ReadOnlySpan` overload, which does the real work, and to which other overloads of `Append` are now forwarded as well. Something similar also happened with the `Insert` family of methods.

Speaking of `Insert`, the overloads that take a primitive were optimized to take advantage of the `ISpanFormattable` interface, eliminating a temporary string allocation most of the time.

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
A note on `AllocateUninitializedArray` usage: this could end up exposing uninitialized memory through the `StringBuilder` instance if multiple threads begin operating simultaneously on the instance and corrupt its internal state. While this is obviously an invalid operation, it is an observable behavioral change from the previous iteration, which would only ever expose zero-inited memory.

See also https://github.com/dotnet/runtime/pull/47186.

**Comment by teo-tsirpanis** [comment]
OK @GrabYourPitchforks, I reverted it.

**Comment by danmoseley** [comment]
@teo-tsirpanis could you please check StringBuilder has decent coverage in dotnet/performance and use the instructions in that repo to generate a before/after table?

**Comment by teo-tsirpanis** [comment]
> and use the instructions in that repo to generate a before/after table

Unfortunately I can't do that; when I had tried to build the runtime, my laptop ran out of space.

I will however take a look at the `StringBuilder` benchmarks soon.

**Comment by teo-tsirpanis** [comment]
Some tests are failing. I will investigate it soon.

**Comment by danmoseley** [comment]
The only relevant failure seems to be Append_Bool_NoSpareCapacity_ThrowsArgumentOutOfRangeException where the parameter name changed. We're usually OK changing the parameter name if there's a reason.

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs [hunk 1]
```diff
@@ -408,14 +408,8 @@ public string ToString(int startIndex, int length)
...
             string result = string.FastAllocateString(length);
-            unsafe
-            {
-                fixed (char* destinationPtr = result)
-                {
-                    this.CopyTo(startIndex, new Span<char>(destinationPtr, length), length);
-                    return result;
-                }
-            }
+            this.CopyTo(startIndex, new Span<char>(ref result.GetRawStringData(), length), length);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:411]:
```suggestion
            CopyTo(startIndex, new Span<char>(ref result.GetRawStringData(), length), length);
```

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs [hunk 2]
```diff
@@ -1205,28 +1170,48 @@ public StringBuilder Append(char[]? value)
         {
             if (value?.Length > 0)
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1171]: Can we get rid of the null and length checks, too?  `value.AsSpan()` will already return an empty span for null and it'll short-circuit for empty.  I doubt we should be optimizing for the null/empty case.

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs [hunk 3]
```diff
@@ -1205,28 +1170,48 @@ public StringBuilder Append(char[]? value)
...
         public StringBuilder Append(ReadOnlySpan<char> value)
         {
-            if (value.Length > 0)
+            // This is where we can check if adding value will put us over m_MaxCapacity.
+            // We are doing the check here to prevent the corruption of the StringBuilder.
+            int newLength = Length + value.Length;
+            if (newLength > m_MaxCapacity || newLength < value.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(value), SR.ArgumentOutOfRange_LengthGreaterThanCapacity);
+            }
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1186]: Shouldn't all of the above be moved to inside the if block?  Doing so shouldn't make the happy path any more expensive, and it'll avoid extra work in the empty span case.

Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1186]: Can this method instead be structured as:
```C#
// This case is so common we want to optimize for it heavily.
if (value.TryCopyTo(m_ChunkChars.AsSpan(m_ChunkLength)))
{
    m_ChunkLength += value.Length;
}
else if (value.Length != 0)
{
    ... // the above argument/state validation
    ... // the contents of the else block below
}
return this;
```
?

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs [hunk 4]
```diff
@@ -1257,53 +1242,41 @@ public StringBuilder Append(ReadOnlySpan<char> value)
 
         #region AppendJoin
 
-        public unsafe StringBuilder AppendJoin(string? separator, params object?[] values)
+        public StringBuilder AppendJoin(string? separator, params object?[] values)
         {
             separator ??= string.Empty;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1247]: This should no longer be necessary.  `separator.AsSpan()` will return an empty span if `separator` is null.

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs [hunk 5]
```diff
@@ -1257,53 +1242,41 @@ public StringBuilder Append(ReadOnlySpan<char> value)
...
-            {
-                return AppendJoinCore(pSeparator, separator.Length, values);
-            }
+            return AppendJoinCore(separator.AsSpan(), values);
         }
 
-        public unsafe StringBuilder AppendJoin<T>(string? separator, IEnumerable<T> values)
+        public StringBuilder AppendJoin<T>(string? separator, IEnumerable<T> values)
         {
             separator ??= string.Empty;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1253]: Ditto

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs [hunk 6]
```diff
@@ -1257,53 +1242,41 @@ public StringBuilder Append(ReadOnlySpan<char> value)
...
-            {
-                return AppendJoinCore(pSeparator, separator.Length, values);
-            }
+            return AppendJoinCore(separator.AsSpan(), values);
         }
 
-        public unsafe StringBuilder AppendJoin(string? separator, params string?[] values)
+        public StringBuilder AppendJoin(string? separator, params string?[] values)
         {
             separator ??= string.Empty;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1259]: Ditto

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs [hunk 7]
```diff
@@ -1376,32 +1349,23 @@ public StringBuilder Insert(int index, string? value)
 
             if (value != null)
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1350]: Same null check comment as earlier.

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs [hunk 8]
```diff
@@ -1412,7 +1376,7 @@ public StringBuilder Insert(int index, char[]? value)
...
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1377]: Ditto

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs [hunk 9]
```diff
@@ -1451,35 +1415,31 @@ public StringBuilder Insert(int index, char[]? value, int startIndex, int charCo
 
             if (charCount > 0)
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1416]: Seems better to remove this check, too.  Seems better to optimize for the != 0 case than to optimize for the == 0 case.

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs (hunk 1 above)
Comment (teo-tsirpanis) [reply src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs]: Done. I also removed all usages of `this.` in the file for consistency.

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs (hunk 3 above)
Comment (teo-tsirpanis) [reply src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1177]: Done (don't know why GitHub doesn't show "Outdated").

**Comment by teo-tsirpanis** [comment]
Feedback was addressed.

**Comment by danmoseley** [comment]
@stephentoub do you have further feedback?

**Comment by teo-tsirpanis** [comment]
Unfortunately I can't provide benchmarks, as explained earlier. I even tried copy-pasting `StringBuilder`'s source file to the benchmark's project but it uses too many internal corelib constructs.

**Comment by teo-tsirpanis** [comment]
Sorry, still can't do it. My machine's SSD runs out of space when I try to build the runtime. I tried to do it in a Codespace but still ran into some issues.

**Comment by stephentoub** [comment]
@teo-tsirpanis, I think much of this PR was obsoleted by https://github.com/dotnet/runtime/pull/64405.  Shall we close it and you can open a new one focusing on just the portions you think are still relevant?  Thanks!

**Pull request closed without merging**